Use quaternion rotation map in BinaryCompactObject domain creator

knelli2 opened the pull request:
Control systems will require us to use the quaternion rotation map.

<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->
When specifying input options for the `BinaryCompactObject` domain creator, replace
```
RotationAboutZAxisMap:
  InitialRotationAngle: 1.0
  InitialAngularVelocity: -0.5
```
with
```
RotationMap:
  InitialAngularVelocity: [0.0, 0.0, -0.5]
```
Note: The initial rotation angle is now always assumed to be zero and cannot be specified by the user.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Use quaternion rotation map in BinaryCompactObject domain creator

knelli2 committed:
Move rot matrix helpers tests to free functions

knelli2 committed:
Decrease abs tol in QuatFoT

This was a very strict tolerance. Decreasing it by this amount
speeds things up and is still accurate enough.

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp:
```diff
@@ -80,6 +83,56 @@ Matrix rotation_matrix_deriv(
 }
 
+DataVector quaternion_from_angles(double theta_x, double theta_y,
+                                  double theta_z) {
```
const args.

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.hpp:
```diff
@@ -26,3 +27,7 @@ Matrix rotation_matrix(double t,
 Matrix rotation_matrix_deriv(
     double t, const domain::FunctionsOfTime::FunctionOfTime& fot);
+
+/// Calculate a unit quaternion from rotation angles about each axis
```
Which set of angles is this?  From the cpp I think they're yaw-pitch-roll (permuted?).

wthrowe commented on src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp:
```diff
@@ -80,6 +83,56 @@ Matrix rotation_matrix_deriv(
+  // being zero.
+  const double qx = 0.5 * sqrt(1.0 + m(0, 0) - m(1, 1) - m(2, 2));
+  const double qy = 0.5 * sqrt(1.0 - m(0, 0) + m(1, 1) - m(2, 2));
+  const double qz = 0.5 * sqrt(1.0 - m(0, 0) - m(1, 1) + m(2, 2));
```
You've got issues with branch cuts here (the correct values can be negative), and switching equations based on roundoff differences usually indicates bad numerical behavior.

Wikipedia [suggests an algorithm](https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion).  I haven't checked it, but it looks sane.

Add tests that would have caught this.

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -473,22 +466,23 @@ class BinaryCompactObject : public DomainCreator<3> {
-  struct RotationAboutZAxisMap {
+  struct RotationMap {
     static constexpr Options::String help = {
         "Options for a time-dependent rotation map about the z axis"};
```
Update.

wthrowe commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -802,7 +809,7 @@ void test_parse_errors() {
 }  // namespace
 
-// [[Timeout, 15]]
+// [[Timeout, 40]]
```
Can we split the slow part of this up?

knelli2 replied:
The "slow" parts of this test are the `test_bbh_time_dependent_factory()` tests which take up about ~90% of this test with these changes. The reason for this is the `QuaternionFunctionOfTime`. Every time it is called, it solves an ODE to get the quaternion, integrating from the last update time to the time requested. However, the "last update time" here is just the initial time. So every time we want the quaternion we are integrating from 0.0 to 7.8 (the last tested time for this test, just as an example) with a tolerance of 1.0e-12. Integrating this far only a couple times is still pretty quick, but I calculated that there are ~83,000 time-dep coord map evaluations in this whole test...so things add up and it gets pretty slow.

One thing I thought about doing was changing the times to check from `{0.0, 4.4, 7.8}` to something like `{0.0, 0.4, 1.1, 1.6}`. This should keep the test under 15 seconds. I was just worried about restricting the times tested to small times.

wthrowe replied:
Yeah, I think that's better.  When the test tolerances are ~1e-10 and the timescales are ~0.2, a time of 1.6 isn't really very small.

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -473,22 +466,23 @@ class BinaryCompactObject : public DomainCreator<3> {
   struct InitialAngularVelocity {
-    using type = double;
+    using type = std::array<double, 3>;
     static constexpr Options::String help = {"The angular velocity."};
```
If I'm reading this all correctly, the input file expects something like Euler angles for the initial angle, but uses an axis-angle formulation for the initial velocity.  Could we do axis-angle for both?  (Trying to do velocities with Euler angles sounds like a mess.)

knelli2 replied:
Yes I think your suggestion of using axis-angle for both is best. Then it's trivial to get the initial quaternion from inputs. Doing that fixes several of the other issues you raised here.

wthrowe reviewed (changes requested):
For the upgrade instructions: what if the values aren't all zero?

knelli2 commented:
Added a fixup and reverted one commit because I didn't need it anymore on account of using only the axis-angle representation.
Edit: Also fixed PR description

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -468,14 +468,14 @@ class BinaryCompactObject : public DomainCreator<3> {
+    using type = double;
     static constexpr Options::String help = {
-        "Rotation angle about each axis at initial time."};
+        "Rotation angle about the initial angular velocity vector."};
```
Why can the initial rotation only be about the angular velocity?

knelli2 replied:
When using the axis-angle representation, the angle is about the axis. The axis here *is* the angular velocity, so the angle is about the angular velocity.

wthrowe replied:
But why do the value and derivative have to share the same axis?

knelli2 replied:
Now I'm questioning if we even need a rotation angle. There are a couple notions of "angle" here and I want to make sure we are talking about the same ones. A unit quaternion represented as q = (cos(theta/2), nhat*sin(theta/2)) will rotate a vector about nhat by an angle theta. Here, nhat is the angular velocity. So "theta" is one notion of angle about the angular velocity. Then there are phi_x, phi_y, phi_z where the x-component of the angular velocity omega_x = dt(phi_x), omega_y = dt(phi_y), etc. This is another notion. I was referring to "theta" in that description, but you seem to be referring to "phi_x" etc. when you say value and derivative.

For initial conditions, we must specify the initial quaternion and initial angular velocity to properly construct things. But since the vector part of the quaternion is just the unit vector in the direction of the angular velocity, we technically only need "theta" to specify the quaternion. After thinking about it, I think we can just set "theta", "phi_x", etc. = 0 always for initial conditions, so the initial quaternion is just (1,0,0,0). (For reference, for SpEC ID, you only specify omega_z and the quaternion is assumed to be (1,0,0,0), implying that theta = 0)

Also, for our initial conditions, are we ever going to start a binary where the angular velocity isn't initially aligned with the z-axis? If we are, then I can leave the InitialAngularVelocity as a vector. But if not, I think we can make it a scalar again and have it just be the z-component of the initial vector.

Thoughts?

wthrowe replied:
I don't think there's a reason to have more freedom here than in SpEC.  Maybe there are cases where a slightly off-axis initial rotation would be better, but I haven't heard about anyone having problems with the SpEC version so any such effect is probably small enough that the control system can quickly compensate.

wthrowe reviewed (commented):
You can squash the existing fixup.

knelli2 commented:
Rebased, squashed in previous fixup, added new fixup.

@wthrowe I decided to remove the rotation angle input option and leave the initial angular velocity as a vector.

wthrowe commented:
Looks fine.  Go ahead and squash again.

kidder reviewed (dismissed):
This LGTM.  @wthrowe this has been squashed.  @geoffrey4444 are you planning on looking at this, or can we merge it?

nilsvu commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -802,7 +809,7 @@ void test_parse_errors() {
 }  // namespace
 
-// [[Timeout, 15]]
+// [[Timeout, 20]]
```
Is it the quaternion ODE integration that slows this down? Is that fine?

nilsvu replied:
I saw your explanation above. So the answer is that in an evolution the ODE integration only happens occasionally and hence doesn't slow things down a lot, but in this test it's called an excessive number of times?

knelli2 replied:
Correct. The quaternion ODE integration is comparable in time to the one in SpEC.

wthrowe reviewed (dismissed):
Code looks fine, but upgrade instructions need to be updated.

Not merging yet to wait for a response to @nilsvu's question.

knelli2 commented:
Updated the upgrade instructions.

nilsvu commented:
@knelli2 looks like you have to increase the timeout further or reduce the number of ODE integrations for the test to pass. gcc-8 Debug is failing.

knelli2 commented:
Rebased and decreased number of ODE integrations by reducing the number of times checked in the test (the range is still the same, just less intermediate times tested).

kidder commented:
ping @geoffrey4444

knelli2 commented:
Fixed conflicts from #3857

kidder commented:
I am going to merge this later today unless I hear that anyone has any remaining issues @wthrowe @nilsvu @geoffrey4444
